packages/driver/src/cy/commands/sessions.ts

```diff
@@ -0,0 +1,822 @@
+    if (!Cypress.config('experimentalSessionSupport')) {
+      // TODO: proper error msg
+      throw new Error('experimentalSessionSupport is not enabled. You must enable the experimentalSessionSupport flag in order to use Cypress session commands')
+    }
+  }
+
+  const sessions = {
+
+    defineSession (options = {} as any) {
+      const sess_state = {
+        id: options.id,
+        cookies: null,
+        localStorage: null,
+        setup: options.setup,
+        hydrated: false,
+        validate: options.validate,
+      }
+
+      setActiveSession({ [sess_state.id]: sess_state })
+
+      return sess_state
+    },
+
+    async clearAllSavedSessions () {
+      clearActiveSessions()
+
+      return Cypress.backend('clear:session', null)
+    },
+
+    async clearCurrentSessionData () {
+      window.localStorage.clear()
+      window.sessionStorage.clear()
+
+      await Promise.all([
+        sessions.clearStorage(),
+        sessions.clearCookies(),
+      ])
+    },
+
+    async setSessionData (data) {
+      const allHtmlOrigins = await getAllHtmlOrigins()
+
+      let _localStorage = data.localStorage || []
+      let _sessionStorage = data.sessionStorage || []
+
+      _.each(allHtmlOrigins, (v) => {
+        if (!_.find(_localStorage, v)) {
+          _localStorage = _localStorage.concat({ origin: v, clear: true })
+        }
+
+        if (!_.find(_sessionStorage, v)) {
+          _sessionStorage = _sessionStorage.concat({ origin: v, clear: true })
+        }
+      })
+
+      await Promise.all([
+        sessions.setStorage({ localStorage: _localStorage, sessionStorage: _sessionStorage }),
+        Cypress.automation('clear:cookies', null),
+      ])
+
+      await sessions.setCookies(data.cookies)
+    },
+
+    getCookies () {
+      return Cypress.automation('get:cookies', {})
+    },
+
+    setCookies (data) {
+      return Cypress.automation('set:cookies', data)
+    },
+
+    async clearCookies () {
+      return Cypress.automation('clear:cookies', await sessions.getCookies())
+    },
+
+    async getCurrentSessionData () {
+      const storage = await sessions.getStorage({ origin: '*' })
+
+      let cookies = [] as any[]
+
+      cookies = await Cypress.automation('get:cookies', {})
+
+      const ses = {
+        ...storage,
+        cookies,
+      }
+
+      return ses
+    },
+
+    getSession (id) {
+      return Cypress.backend('get:session', id)
+    },
+
+    /**
+     * 1) if we only need current_origin localStorage, access sync
+     * 2) if cross-origin http, we need to load in iframe from our proxy that will intercept all http reqs at /__cypress/automation/*
+     *      and postMessage() the localStorage value to us
+     * 3) if cross-origin https, since we pass-thru https conntections in the proxy, we need to
+     *      send a message telling our proxy server to intercept the next req to the https domain,
+     *      then follow 2)
+     */
+    async getStorage (options = {}) {
```

Comment:
nit: this function is >100 lines long, i get that it's a complicated procedure, but it would be easier to grok if it was broken up into smaller named functions

Reply:
addressed